docs/fsharp-cheatsheet.md

```diff
@@ -1,523 +1,630 @@
 
-    int -> string                       // this defines a function that receives an integer; returns a string
-    int -> int -> string                // two integer inputs; returns a string
-    unit -> string                      // unit; returns a string
-    string -> unit                      // accepts a string; no return
-    (int * string) -> string -> string  // a tuple of int and string, and a string inputs; returns a string
+```fsharp
+int -> string                       // this defines a function that receives an integer; returns a string
+int -> int -> string                // two integer inputs; returns a string
+unit -> string                      // unit; returns a string
+string -> unit                      // accepts a string; no return
+(int * string) -> string -> string  // a tuple of int and string, and a string inputs; returns a string
+```
 
 Most of the time, the compiler can determine the type of a parameter, but there are cases may you wish to be explicit or the compiler needs a hand.
 Here is a function with a signature `string -> char -> int` and the input and return types are explicit:
 
-    let countWordsStartingWithLetter (theString: string) (theLetter: char) : int =
-        theString.Split ' '
-        |> Seq.where (fun (word: string) -> word.StartsWith theLetter)  // explicit typing in a lambda
-        |> Seq.length
+```fsharp
+let countWordsStartingWithLetter (theString: string) (theLetter: char) : int =
+    theString.Split ' '
+    |> Seq.where (fun (word: string) -> word.StartsWith theLetter)  // explicit typing in a lambda
+    |> Seq.length
+```
 
 Examples of functions that take [`unit`](#functions-unit-type) as arguments and return different [Collection](#collections) types.
 
-    let getList (): int list = ...  // unit -> int list
-    let getArray (): int[] = ...
-    let getSeq (): seq<int> = ...
+```fsharp
+let getList (): int list = ...  // unit -> int list
+let getArray (): int[] = ...
+let getSeq (): seq<int> = ...
+```
 
 A complex declaration with an [Anonymous Record](#data-types-anonymous-records):
 
-    let anonRecordFunc (record: {| Count: int; LeftAndRight: bigint * bigint |}) =
-        ...
+```fsharp
+let anonRecordFunc (record: {| Count: int; LeftAndRight: bigint * bigint |}) =
+    ...
+```
 
 <div id="functions-recursive"></div>
 
-### Recursive
+## Recursive Functions
 
 The `rec` keyword is used together with the `let` keyword to define a recursive function:
 
-    let rec fact x =
-        if x < 1 then 1
-        else x * fact (x - 1)
+```fsharp
+let rec fact x =
+    if x < 1 then 1
+    else x * fact (x - 1)
+```
+
+### TailCallAttribute
+
+In _tail recursive_ functions, the recursive call is the final operation in the function, with its result directly returned without further computation. This pattern allows the compiler to optimize memory usage by reusing the current stack frame instead of allocating a new one for each call.
```

Comment:
```suggestion
In _tail recursive_ functions, the recursive call is the final operation in the function, with its result directly returned without a nested function call (and the stack usage that implies). This pattern allows the compiler to instead generate a loop equivalent of the nested invocation by reusing the current stack frame instead of allocating a new one for each call.
```